core/src/main/scala/zio/config/ConfigSource.scala

```diff
@@ -16,38 +16,30 @@ import zio.Task
 
 import scala.annotation.tailrec
 
-sealed trait ConfigSource[K, V] { self =>
-  def sourceDescription: Set[String]
-
-  def getConfigValue(path: List[K]): PropertyTree[K, V]
-
-  final def orElse(that: => ConfigSource[K, V]): ConfigSource[K, V] = new MergedConfigSource(self, that)
-
-  final def <>(that: => ConfigSource[K, V]): ConfigSource[K, V] = self orElse that
-}
-
-final case class TreeConfigSource[K, V](tree: PropertyTree[K, V], sourceDescription: Set[String])
-    extends ConfigSource[K, V] {
-  def getConfigValue(path: List[K]): PropertyTree[K, V] = tree.getPath(path)
-}
-final class MergedConfigSource[K, V](_left: => ConfigSource[K, V], _right: => ConfigSource[K, V])
-    extends ConfigSource[K, V] {
-  private lazy val left  = _left
-  private lazy val right = _right
-
-  def getConfigValue(path: List[K]): PropertyTree[K, V] =
-    left.getConfigValue(path).getOrElse(right.getConfigValue(path))
+final case class ConfigSource[K, V](
+  sourceDescription: Set[String],
+  getConfigValue: List[K] => PropertyTree[K, V]
+) { self =>
+
+  def orElse(that: => ConfigSource[K, V]): ConfigSource[K, V] =
+    ConfigSource(
+      self.sourceDescription ++ that.sourceDescription,
+      path => self.getConfigValue(path).getOrElse(that.getConfigValue(path))
+    )
 
-  lazy val sourceDescription: Set[String] = left.sourceDescription ++ right.sourceDescription
+  def <>(that: => ConfigSource[K, V]): ConfigSource[K, V] = self orElse that
 }
 
 object ConfigSource {
   private[config] val SystemEnvironment    = "system environment"
   private[config] val SystemProperties     = "system properties"
   private[config] val CommandLineArguments = "command line arguments"
 
-  def empty[K, V]: ConfigSource[K, V]                                       = TreeConfigSource(PropertyTree.empty, Set.empty)
-  def apply[K, V](tree: PropertyTree[K, V], sourceDescription: Set[String]) = TreeConfigSource(tree, sourceDescription)
+  def empty[K, V]: ConfigSource[K, V] =
+    ConfigSource(Set.empty, _ => PropertyTree.empty)
+
+  def mk[K, V](tree: PropertyTree[K, V], descriptions: Set[String]): ConfigSource[K, V] =
```

Comment:
This should probably be renamed, `fromPropertyTree` and sorted alphabetically.

Reply:
There is already `fromPropertyTree`. this was a carry-over from previous merges. I have removed the function as such and reused `fromPropertyTree`